### PR TITLE
feat: Added pprof and variable in makefile

### LINKED ISSUE #1059: Publish a debug image with reach stable/latest release.
## Feature Request

**Short Description**

We want to have a debuggable release (pprof enabled) along side the nomal release.

**Describe the solution you'd like**

We would like to enable pprof conditionally when compiling KubeArmor binary.
e.g:  `ENABLE_PPROF=true make run` to enable pprof and run KubeArmor.

We also would like to push for each release(stable or latest) a second image with pprof enabled.
The image should be named as follow: `<version/state>-debug`
eg: `latest-debug` `stable-debug` `v0.8-debug`

### PR DESCRIPTION
**Pull request by mdsahiloss**
Signed-off-by: MdSahil-oss <Mohdssahil1@gmail.com>

**Purpose of PR?**:

Fixes #1059

**Does this PR introduce a breaking change?**
no
**If the changes in this PR are manually verified, list down the scenarios covered:**:

**Additional information for reviewer?** :
_Mention if this PR is part of any design or a continuation of previous PRs_

**Checklist:**
- [x] Bug fix. Fixes #1059
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update
- [ ] PR Title follows the convention of  `<type>(<scope>): <subject>`
- [ ] Commit has unit tests
- [ ] Commit has integration tests

### CODE REVIEW

**Comment by mdsahiloss** [comment]
@achrefbensaad I've added `pprof` support. can you please review it ? also Let me know how can I perform testing locally using minikube? confused about it.

**Comment by achrefbensaad** [comment]
Hi @MdSahil-oss,  Sure will do. Yes you can use minikube to test locally.

**Comment by mdsahiloss** [comment]
@achrefbensaad Thanks for letting me know but what's the procedure to test my changes using minikube ?

**Comment by achrefbensaad** [comment]
you can use this method.
1- use `build_kubearmor.sh` to generate a docker image, make sure to set the `REPO` env variable with the entended image name or it will create an image with the name `kubearmor/kubearmor`.
2- push the image to a container registry
3- using karmor cli, run `karmor install -i <your-image-name>`.

**Comment by mdsahiloss** [comment]
@achrefbensaad I tested `pprof` It's working fine according to me in this PR. Please review the flags that I've added

**Comment by mdsahiloss** [comment]
And confused about how to generate `pprof` enabled images because Dockerfile is just compiling the code with `make` command. Should I create a new Dockerfile which will create `pprof` enabled image when executing `build_kubearmor.sh` command

**Comment by achrefbensaad** [comment]
Hi @MdSahil-oss , there is no need to create a new dockerfile, you cna modify the ci to build and push two images, one with pprof enabled and one without. Also can you please implement the rest of API's mentioned on the issue.

**Comment by mdsahiloss** [comment]
@achrefbensaad APIs like:
pprof.Handler("heap") -> /debug/pprof/heap
pprof.Handler("block") -> /debug/pprof/block
pprof.Handler("goroutine") -> /debug/pprof/goroutine
pprof.Handler("threadcreate") ->/debug/pprof/threadcreate
By default accessible that's why I didn't implement these.

**Comment by achrefbensaad** [comment]
Ah ok, Ill test the PR. In the meantime, can you please rebase on top of main to resolve the gosec error.

**Comment by mdsahiloss** [comment]
@achrefbensaad The flow of generating `pprof` enabled images I can see is:
- First CI will trigger `build_kubearmor.sh` with a tag `<state/version>-debug` as an argument.
- By checking tag as an argument conditionally `build_kubearmor.sh` will call dockerfiles to create images.
- But if we want to create `pprof` enabled images then `ENABLE_PPROF` must be set to `true` by `ENABLE_PPROF=true make` command
As far as I can see dockerfiles is the only file which runs `make` command so i don't see anyother way to create pprof `enabled` images without creating a new dockerfile. If you have anyother method to do so then let me know will be glad to hear it :)

**Comment by achrefbensaad** [comment]
you can pass the value of `ENABLE_PPROF` to the dockerfile and then pass it down to the makefile using env vars.
This is a link that shows how this can be achieved https://stackoverflow.com/questions/30494050/how-do-i-pass-environment-variables-to-docker-containers

**Comment by mdsahiloss** [comment]
@achrefbensaad this PR needs to be reviewed now, Let me know if you find anything to be changed.

File: .github/workflows/ci-latest-release.yml [hunk 1]
```diff
@@ -121,7 +122,9 @@ jobs:
           value=`cat STABLE-RELEASE`
           if [ ${{ github.ref }} == refs/heads/$value ]; then
             docker pull docker.io/kubearmor/kubearmor:$value
+            docker pull docker.io/kubearmor/kubearmor:$value-debug
             docker image tag kubearmor/kubearmor:$value kubearmor/kubearmor:stable
+            docker image tag kubearmor/kubearmor:$value-debug kubearmor/kubearmor:stable-debug
```
Comment (achrefbensaad) [inline .github/workflows/ci-latest-release.yml]: We should add a new step rather then changing the existant step.

File: Dockerfile [hunk 2]
```diff
@@ -14,6 +14,9 @@ COPY . .
 
 WORKDIR /usr/src/KubeArmor/KubeArmor
 
+ARG ENABLE_PPROF
+ENV ENABLE_PPROF=$ENABLE_PPROF
+
 RUN go install github.com/golang/protobuf/protoc-gen-go@latest
 RUN make
```
Comment (achrefbensaad) [inline Dockerfile]: you need to pass the variable to the make file.

File: KubeArmor/build/kubearmor-test-docker.yaml [hunk 3]
```diff
@@ -152,6 +152,76 @@ spec:
         - mountPath: /var/lib/docker
           name: docker-storage-path
           readOnly: true
+      - args:
```
Comment (achrefbensaad) [inline KubeArmor/build/kubearmor-test-docker.yaml]: No need to change this file.

File: KubeArmor/build/build_kubearmor.sh [hunk 4]
```diff
@@ -45,7 +45,7 @@ unset LABEL
 # build a kubearmor image
 DTAG="-t $REPO:$VERSION"
 echo "[INFO] Building $DTAG"
-cd $ARMOR_HOME/..; docker build $DTAG -f Dockerfile --target kubearmor . $LABEL
+cd $ARMOR_HOME/..; docker build $DTAG -f Dockerfile --target kubearmor . $LABEL -e ENABLE_PPROF=false
```
Comment (achrefbensaad) [inline KubeArmor/build/build_kubearmor.sh:53]: the value should be handed down from the CI to this script, to the docker file , to the make file

**Review by achrefbensaad — CHANGES_REQUESTED** [review]

**Comment by mdsahiloss** [comment]
@achrefbensaad I've updated the PR.

**Comment by achrefbensaad** [comment]
Lets start by fixing the points above, then we will fix the other parts

**Comment by mdsahiloss** [comment]
@achrefbensaad I think I have fixed all the points that you mentioned :)

**Comment by achrefbensaad** [comment]
Hi @MdSahil-oss , yes you have addressed the points I raised earlier, but some new issues were introduced also, please read the reviews above my comment.
Also. The CI should not generate only debug images (PPROF_ENABLED=true), we need to keep the normal release also (PPROF_ENABLED=false).

**Comment by mdsahiloss** [comment]
@achrefbensaad I think I have resolved all the reviews you did above.

File: KubeArmor/build/build_kubearmor.sh [hunk 5]
```diff
@@ -26,12 +26,18 @@ cd $ARMOR_HOME/build
 pwd
 
 VERSION=latest
+ENABLE_PPROF=true
```
Comment (Ankurk99) [inline KubeArmor/build/build_kubearmor.sh]: why are you setting it to true here when it's already done in the ci manifest?

**Review by Ankurk99 — CHANGES_REQUESTED** [review]
@MdSahil-oss Please address the unresolved commits.

**Comment by mdsahiloss** [comment]
@achrefbensaad @Ankurk99  Now the value of `ENABLE_PPROF` is handed down from the CI to this script, to the docker file , to the make file as per your suggestion

**Comment by mdsahiloss** [comment]
And I kept `ENABLE_PPROF=false` for pprof disabled images.

**Review by daemon1024 — CHANGES_REQUESTED** [review]
@MdSahil-oss Sorry that we didn't follow up on this PR. The changes LGTM, can you please rebase your PR to resolve conflicts. Thank You.

**Comment by PrimalPimmy** [comment]
Fixed in #1495 , closing this PR

**Pull request closed without merging**